core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1327,69 +1327,337 @@ class ReassignedPartitionsIsrChangeListener(protected val controller: KafkaContr
+          controllerContext.partitionsOnBroker(id)
+            .map(topicAndPartition => (topicAndPartition, controllerContext.partitionReplicaAssignment(topicAndPartition).size))
+
+      allPartitionsAndReplicationFactorOnBroker.foreach {
+        case(topicAndPartition, replicationFactor) =>
+          controllerContext.partitionLeadershipInfo.get(topicAndPartition).foreach { currLeaderIsrAndControllerEpoch =>
+            if (replicationFactor > 1) {
+              if (currLeaderIsrAndControllerEpoch.leaderAndIsr.leader == id) {
+                // If the broker leads the topic partition, transition the leader and update isr. Updates zk and
+                // notifies all affected brokers
+                partitionStateMachine.handleStateChanges(Set(topicAndPartition), OnlinePartition,
+                  controlledShutdownPartitionLeaderSelector)
+              } else {
+                // Stop the replica first. The state change below initiates ZK changes which should take some time
+                // before which the stop replica request should be completed (in most cases)
+                try {
+                  brokerRequestBatch.newBatch()
+                  brokerRequestBatch.addStopReplicaRequestForBrokers(Seq(id), topicAndPartition.topic,
+                    topicAndPartition.partition, deletePartition = false)
+                  brokerRequestBatch.sendRequestsToBrokers(epoch)
+                } catch {
+                  case e : IllegalStateException => {
+                    // Resign if the controller is in an illegal state
+                    error("Forcing the controller to resign")
+                    brokerRequestBatch.clear()
+                    triggerControllerMove()
+
+                    throw e
+                  }
+                }
+                // If the broker is a follower, updates the isr in ZK and notifies the current leader
+                replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
+                  topicAndPartition.partition, id)), OfflineReplica)
+              }
+            }
           }
-          topicAndPartitions
-        case None =>
-          error("Invalid topic and partition JSON: " + jsonOpt.get + " in ZK: " + changeZnode)
-          Set.empty
       }
-    } else {
-      Set.empty
+      def replicatedPartitionsBrokerLeads() = {
+        trace("All leaders = " + controllerContext.partitionLeadershipInfo.mkString(","))
+        controllerContext.partitionLeadershipInfo.filter {
+          case (topicAndPartition, leaderIsrAndControllerEpoch) =>
+            leaderIsrAndControllerEpoch.leaderAndIsr.leader == id && controllerContext.partitionReplicaAssignment(topicAndPartition).size > 1
+        }.keys
+      }
+      replicatedPartitionsBrokerLeads().toSet
     }
   }
+
+  case class TopicDeletionStopReplicaResult(stopReplicaResponseObj: AbstractResponse, replicaId: Int) extends ControllerEvent {
```

Comment:
None of the other events end in "Event" so to stay consistent, I'd rather keep it as is.